### ISSUE TITLE
Deprecate in favor of common ocaml-posix

### ISSUE THREAD

**Issue by toots** [mwweissmann/ocaml-posix-time#1]
Hi,

Following a conversation with the `ocaml-ctypes` maintainer here: https://github.com/ocamllabs/ocaml-ctypes/pull/620 I have worked on a consolidated [ocaml-posix](https://github.com/savonet/ocaml-posix) module.

This module aims at giving a unified structure to all existing POSIX bindings, with support for both high-level consumer API as well as low-level ctypes APIs.

Also, these modules take advantage of the most recent `dune` support which, in particular, allows for powerful compilation configuration, including multi-layered builds and config tests. which makes it possible to write all these bindings without writing a single C file. The gist of this scaffolding is described here: https://medium.com/@romain.beauxis/advanced-c-binding-using-ocaml-ctypes-and-dune-cc3f4cbab302

The `posix-time` module included in `ocaml-posix` provides the exact same API as this module. Considering that this module has not been updated for a while and that the new one would be a drop-in, consolidated replacement, would you be okay to deprecate this module to be replaced by the new one?

It is, of course, my intent to give all previous maintainer commit access to the new repository so the work can continue jointly there.

Let me know what you think!
Romain

**Comment by mwweissmann** [comment]
Hi Romain,
the posix-time module is only just a helper library for other libraries, most importantly ocaml-socketcan -- I don't care much for the "library" itself. I don't currently see much benefit in rewriting the dependants with ctypes; I care most for stability and performance (on low end devices) especially in socketcan, so I am very skeptical about using ffi for that.

**Comment by toots** [comment]
Hi @mwweissmann & thanks for your response.

I totally understand the issue with performances. However, what I am proposing here is a native binding, not a dynamically loaded API via `libffi`. I believe the performances should be on-par in this case.

I realize now that your library also ships a C header that is used in `socketcan`. That's a little odd, in my opinion since I don't think that this header would be of much use to OCaml users of `ocaml-posix-time`.

The reason for re-implementing using ctypes, however, follows a similar pattern: surfacing _native_ ctypes modules for each structures allows developers who want to build new bindings using these common POSIX types to be able to readily use them in their own modules.

I developed that idea and pattern while working on a binding for libsrt. The stubs here: https://github.com/savonet/ocaml-srt/blob/master/src/stubs/srt_stubs.ml can just refer to `Posix_socket` and use the common socket type to export the libsrt functions using it.

As I see it, expanding this pattern would be highly beneficial to help building better bindings for the community as a whole. Considering that time types are pretty common C types to handle, it would be great to be able to surface them in the same way.

Let me know what you think. In my opinion, moving the header file from this module to `ocaml-socketcan` and allowing the common `ocaml-posix` to replace it would be a great solution fitting everybody's needs here. I understand if you feel differently, though.

**Comment by mwweissmann** [comment]
The posix-time lib is itself only an interface that I use in multiple C bindings. What I require is a stable memory layout, so I can safely safely assume the memory layout of the OCaml value in my _classical_ native bindings in C. Thats why the header file is with posix-time as this is the place that guarantees the memory layout.
To deprecate posix-time I would first need to rewrite my libraries that rely on it. And I also absolutely must have that stable memory layout as I am not willing to rewrite socketcan with ctypes. Or can ctypes generate the C headers? Functions to access and create the OCaml values? If ocaml-posix would ship with those headers there would be at least the possibility to make modifications in ocaml-posix later w/o breaking the interface with classic C bindings.

**Comment by toots** [comment]
Yes, `ctypes` can do that. The way it works when dealing with structures is as follows:
* During a first pass, `ctypes` calculates the `offset` and `alignment` of each entry in the C structure, via a compiled C program outputing values calculated using `sizeof` and `offsetof`.

* Then, it is able to use these informations to allocate valid C structures that can be passed down to C bindings

Eventually, `ctypes` also generates the C stubs code to call the underlying function bindings, for instance:
```C
value posix_time_3_clock_gettime(value x18, value x17)
{
   CAMLparam2 (x18, x17);;
   uint32_t x19 = Uint32_val(x18);
   struct timespec* x22 = CTYPES_ADDR_OF_FATPTR(x17);
   caml_release_runtime_system();
   int x23 = clock_gettime(x19, x22);
   caml_acquire_runtime_system();
   value x24 = Val_long(x23);
   CAMLreturnT(value, x24);
}
```

Conversion between C and OCaml data structures are handled in the OCaml code. Typically, it looks like this, going from OCaml to C:
```ocaml
  let from_ocaml tm =
    let c_tm = make Types.Tm.t in
    setf c_tm Types.Tm.tm_sec tm.tm_sec;
    setf c_tm Types.Tm.tm_min tm.tm_min;
    setf c_tm Types.Tm.tm_hour tm.tm_hour;
    setf c_tm Types.Tm.tm_mday tm.tm_mday;
    setf c_tm Types.Tm.tm_mon tm.tm_mon;
    setf c_tm Types.Tm.tm_year tm.tm_year;
    setf c_tm Types.Tm.tm_wday tm.tm_wday;
    setf c_tm Types.Tm.tm_yday tm.tm_yday;
    setf c_tm Types.Tm.tm_isdst tm.tm_isdst;
    c_tm
```
and, conversely:
```ocaml
  let to_ocaml c_tm =
    let get f = getf c_tm f in
    {
      tm_sec = get Types.Tm.tm_sec;
      tm_min = get Types.Tm.tm_min;
      tm_hour = get Types.Tm.tm_hour;
      tm_mday = get Types.Tm.tm_mday;
      tm_mon = get Types.Tm.tm_mon;
      tm_year = get Types.Tm.tm_year;
      tm_wday = get Types.Tm.tm_wday;
      tm_yday = get Types.Tm.tm_yday;
      tm_isdst = get Types.Tm.tm_isdst;
    }
```

This gives a stable memory representation lifted up to the OCaml layer.

Now, this being said, by no means was I suggesting that you rewrite your application. It just seemed to me that considering that the OCaml API exported by the rewritten module is the same as your module, moving the C header from the old module to `socketcan` would make `ocaml-socketcan` work with the new module without having to change any code at all on your end.

Let me know if you have any suggestion to move forward. I think we have 3 possible soutions:
1. Rename the new module and avoid conflicts
2. Implement `ocaml-socketcan` using `ctypes`
3. Move `posix-time.h` from your `posix-type` module to `ocaml-socketcan` and switch`ocaml-socketcan` to the new `posix-time` (best solution in my opinion)

Let me know if you have a preference. I'm happy to do the work for 2. and 3.

Thanks!

**Comment by toots** [comment]
Hi! Just wanted to ping you one last time about this. I'm ready to submit a PR to ocaml-ctypes to take over the posix bindings there. If I don't hear back, I'll just rename my module `posix-time2`...
